### PR TITLE
add per-stream biasing to RNN-T, TDT streaming beam search

### PR DESCRIPTION
**Pull request by lilithgrigoryan**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

1. **TDT hypothesis pruning:** This change updates how TDT models prune hypotheses during decoding.

**Before:**  
Hypotheses were selected using top-k over the full joint space: ```(V + 1) × num_durations```

This means the decoder selected the top-k combinations of token and duration jointly.

**After:**
Hypotheses are pruned in two steps:

top-k over ```(V + 1)``` tokens
top-1 over ```num_durations```

This makes beam transcripts more diverse and improves word-boosting quality for TDT models

2. Adds per-stream biasing support to RNN-T and TDT streaming beam search for ```malsd_batch``` decoding algorithm

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

```bash
python examples/asr/asr_chunked_inference/rnnt/speech_to_text_streaming_infer_rnnt.py \
  pretrained_name=nvidia/stt_en_fastconformer_transducer_large \
  dataset_manifest=/home/lgrigoryan/data/en/nvtalks_dataset/nemo_manifest_test_boost_gt.jsonl \
  batch_size=128 \
  output_filename=work/streaming_beam_search_pr/multibiasing/transducer_malsd_streaming_simfalse_biastrue_cudafalse.json \
  allow_mps=true \
  compute_dtype=bfloat16 \
  use_per_stream_biasing=true \
  calculate_rtfx=true \
  left_context_secs=10 \
  chunk_secs=2 \
  right_context_secs=2 \
  decoding.strategy=malsd_batch \
  decoding.beam.beam_size=4 \
  decoding.beam.allow_cuda_graphs=false
```

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/rnnt_malsd_batched_computer.py [hunk 1]
```diff
@@ -278,24 +287,86 @@ def __init__(
...
+    def _fusion_scores_alpha_sum(
+        self,
+        batch_size: int,
+        multi_biasing_ids: Optional[torch.Tensor] = None,
+    ) -> Union[float, torch.Tensor]:
+        """Alpha sum for blank LM weighting; per-stream biasing returns [batch_size, 1]."""
+        lm_alpha = sum(self.fusion_models_alpha)
+        if not self.per_stream_biasing_enabled or multi_biasing_ids is None:
+            return lm_alpha
+        ids = multi_biasing_ids[:batch_size]
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/rnnt_malsd_batched_computer.py]: Let's move this logic to `GPUBiasingMultiModel.get_alphas(model_ids)` (which will also guarantee to return `0` weight for `-1` ids)
Note that `clamp` is not necessary, since `-1` works with tensors (replacing it with `0` still requires at least one model to exist; you are replacing scores for `-1` ids anyway)

File: nemo/collections/asr/parts/submodules/tdt_malsd_batched_computer.py [hunk 2]
```diff
@@ -291,24 +301,86 @@ def __init__(
...
+    def _fusion_scores_alpha_sum(
+        self,
+        batch_size: int,
+        multi_biasing_ids: Optional[torch.Tensor] = None,
+    ) -> Union[float, torch.Tensor]:
+        """Alpha sum for blank LM weighting; per-stream biasing returns [batch_size, 1]."""
+        lm_alpha = sum(self.fusion_models_alpha)
+        if not self.per_stream_biasing_enabled or multi_biasing_ids is None:
+            return lm_alpha
+        ids = multi_biasing_ids[:batch_size]
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_malsd_batched_computer.py]: Same as above

**Review by artbataev — COMMENTED** [review]
Please move the logic to `GPUBiasingMultiModel.get_alphas` or smth like this.
Everything else looks good. 
Also, there is a failing test on CI - please fix.

File: nemo/collections/asr/parts/submodules/tdt_malsd_batched_computer.py (hunk 2 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/tdt_malsd_batched_computer.py]: Done thanks

File: nemo/collections/asr/parts/submodules/tdt_malsd_batched_computer.py [hunk 3]
```diff
@@ -693,131 +748,141 @@ def modified_alsd_torch(
...
                 )
-
                 blank_logprob = log_probs[..., -1]
                 non_blank_logprob = torch.log1p(-torch.clamp(torch.exp(blank_logprob), max=1.0 - eps))
-
                 masked_labels = torch.where(labels_top_k == self._blank_index, 0, labels_top_k)
                 log_probs_top_k = torch.where(
                     labels_top_k == self._blank_index,
-                    log_probs_top_k * (1 + fusion_scores_sum_alpha),
+                    log_probs_top_k * (1 + fusion_scores_alpha_sum),
```
[the remark this answers is not in the document]
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/tdt_malsd_batched_computer.py:817]: Not a bug

File: tests/collections/asr/decoding/test_streaming_beam_decoding.py [hunk 4]
```diff
@@ -418,3 +555,46 @@ def test_malsd_streaming_batched_state_with_word_boosting(
...
     )
     assert ref_transcripts == streaming_transcripts
+
+
+@pytest.mark.with_downloads
+@pytest.mark.parametrize("device,cuda_graphs_mode", DEVICE_PARAM_MATRIX)
+@pytest.mark.parametrize("is_tdt", [False, True])
+@pytest.mark.parametrize("chunk_size", [1])
+@pytest.mark.parametrize("batch_size", [4])
+@pytest.mark.parametrize("beam_size", [4])
```
[the remark this answers is not in the document]
Comment (lilithgrigoryan) [reply tests/collections/asr/decoding/test_streaming_beam_decoding.py:565]: Not a bug

File: nemo/collections/asr/parts/submodules/rnnt_malsd_batched_computer.py [hunk 5]
```diff
@@ -705,7 +793,12 @@ def topk_fusion_model(self, fusion_scores_list, log_probs, eps=1e-2):
                     labels_top_k == self._blank_index,
                     log_probs_top_k * (1 + fusion_scores_alpha_sum),
```
[the remark this answers is not in the document]
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/rnnt_malsd_batched_computer.py:791]: Not a bug

File: nemo/collections/asr/parts/submodules/rnnt_malsd_batched_computer.py (hunk 1 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/rnnt_malsd_batched_computer.py]: Done thanks

**Review by artbataev — APPROVED** [review]
Thanks!
